deployment/promotion/promote.py

```diff
@@ -9,396 +9,157 @@
-
-    print("Starting promotion")
-    build_location = upload_build(promotion['build'])
-    if not dest['registered_model_name']:
-        model = create_model(dest_auth, promotion['model'], promotion['artifacts'])
-    else:
-        models = get_registered_models_by_name(dest_auth, dest['registered_model_name'])
-        if len(models) > 1:
-            print("WARNING: Multiple registered models with name '%s' found, using first one with id '%s'" % (dest['registered_model_name'], models[0]["id"]))
-        elif len(models) == 0:
-            print("ERROR: Registered model with name '%s' not found" % dest['registered_model_name'])
-            return
-        model = models[0]
-        print("Using existing registered model '%s'" % model['name'])
-    model_version = create_model_version(dest_auth, promotion['model_version'], model)
-
-    artifacts_and_model = promotion['artifacts']
-    artifacts_and_model.append(model_version['model'])
-    artifact_paths = upload_artifacts(dest_auth, model_version['id'], artifacts_and_model)
-
-    # Standard RMVs will have artifact path 'model' while ER->RMVs will have artifact path 'model.pkl'
-    model_artifact = model_version['model']
-    if 'model' in artifact_paths.keys():
-        model_artifact['path'] = artifact_paths['model']
-    else:
-        model_artifact['path'] = artifact_paths['model.pkl']
-    patch_model(dest_auth, model['id'], model_version['id'], model_artifact)
-
-    dest_build = create_build(dest_auth, model_version['id'], build_location, promotion['build'])
-    update_build_message(dest_auth, promotion['build'], dest_build)
-    return dest_build
-
-
-def promote(_config):
-    promotion = get_promotion_data(_config)
-    promoted_build = create_promotion(_config, promotion)
-    print("Promoted build %d created." % promoted_build['id'])
-
-
-promote(config)
+    print("Fetching or creating destination registered model %s" % source_model.name)
+    dest_model = dest_client.get_or_create_registered_model(name=source_model.name)
+
+# Copy information from the source model to the destination model
+# TODO: add readme, description to the client
+print("Copying info from source model to destination model")
+dest_model.add_labels([v for v in source_model.get_labels()])
+
+# Create the basic model version
+# TODO: add readme, description to the client
+# TODO: check return types
+print("Copying info from source model version to destination model version")
+dest_model_version = dest_model.create_version(
+    name=source_model_version.name,
+    labels=[v for v in source_model_version.get_labels()],
+    attrs={k: v for k, v in source_model_version.get_attributes().items()},
+)
+
+for key in source_model_version.get_artifact_keys():
+    artifact = source_model_version.get_artifact(key)
+    dest_model_version.log_artifact(key, artifact, overwrite=True)
+
+model = source_model_version.get_model()
+dest_model_version.log_model(model, custom_modules=[], overwrite=True)
```

Comment:
does `custom_modules=[]` mean the script doesn't work with custom modules? Should it?

Reply:
We know the customer uses our Open API / Model Schema feature (https://docs.verta.ai/verta/deployment/guides/model-schema-specification), which requires `code_dependencies`. We'll need to carry that over from source to destination also.

Reply:
All annotations and artifacts are copied from source to destination. I believe that should cover everything. Am I missing something?

Reply:
okay I looked at the `log_model` code and I think this will be fine. Could be worth testing this script for an Open Api model though, just for certainty.